### PR TITLE
Add parameters for fog end depth and use alpha as density.

### PR DESCRIPTION
**Pull request by dlasalle**
This implements fog density via the alpha channel to implement #3806, and adds the option to specify a fog end depth separate from the camera far plane (requested in #12454).
* This allows a fog saturation of less than 100%, if alpha is less than 1.0 and the fog depth end is less than the camera far plane.
* Spreading out fog across multiple cameras in layered rendering (so that fog only reaches 50% saturation in the first camera, and goes from 50% to 100% in the second camera), by having a fog end depth greater than the camera far plane.

Implemented for both GLES2 and GLES3:
gles2
gles3

### CODE REVIEW

**Comment by Calinou** [comment]
Thanks a lot for implementing this :smiley:

Out of curiosity, why does the fog look brighter in the GLES3 screenshot compared to GLES2?

**Comment by reduz** [comment]
This looks good, I would simplify it though and make it work like the shadowmap range. Just add a  "max_depth" parameter and leave it as 0. If >0, then that is the limit.

File: drivers/gles3/rasterizer_scene_gles3.h [hunk 1]
```diff
@@ -518,6 +521,7 @@ class RasterizerSceneGLES3 : public RasterizerScene {
 			fog_depth_enabled = true;
 
 			fog_depth_begin = 10;
+			fog_depth_end = 100;
```
Comment (reduz) [inline drivers/gles3/rasterizer_scene_gles3.h]: Should likely be 0 by default to avoid breaking compatibility.

**Review by reduz — CHANGES_REQUESTED** [review]

File: drivers/gles2/rasterizer_scene_gles2.cpp [hunk 2]
```diff
@@ -2054,7 +2055,11 @@ void RasterizerSceneGLES2::_render_render_list(RenderList::Element **p_elements,
 	if (p_env && !p_shadow && p_env->fog_enabled && (p_env->fog_depth_enabled || p_env->fog_height_enabled)) {
 		state.scene_shader.set_conditional(SceneShaderGLES2::FOG_DEPTH_ENABLED, p_env->fog_depth_enabled);
 		state.scene_shader.set_conditional(SceneShaderGLES2::FOG_HEIGHT_ENABLED, p_env->fog_height_enabled);
-		fog_max_distance = p_projection.get_z_far();
+		if (p_env->fog_depth_end > 0) {
```
Comment (akien-mga) [inline drivers/gles2/rasterizer_scene_gles2.cpp]: I wonder if this should use an epsilon (`CMP_EPSILON` if it's exposed here, or just `0.00001`) to take into account potential floating point precision issues?

Comment (reduz) [reply drivers/gles2/rasterizer_scene_gles2.cpp]: No, this is fine. The editor will always put zero in here so it is reliable to check it like this.

**Pull request merged**